src/baldr/graphreader.cc

```diff
@@ -37,31 +37,55 @@ tile_gone_error_t::tile_gone_error_t(std::string prefix, baldr::GraphId edgeid)
 
 GraphReader::tile_extract_t::tile_extract_t(const boost::property_tree::ptree& pt) {
   // if you really meant to load it
-  if (pt.get_optional<std::string>("tile_extract")) {
+  auto extract_fp = pt.get_optional<std::string>("tile_extract");
+  if (extract_fp) {
     try {
-      // load the tar
-      archive.reset(new midgard::tar(pt.get<std::string>("tile_extract")));
-      // map files to graph ids
-      for (auto& c : archive->contents) {
-        try {
-          auto id = GraphTile::GetTileId(c.first);
-          tiles[id] = std::make_pair(const_cast<char*>(c.second.first), c.second.second);
-        } catch (...) {
-          // It's possible to put non-tile files inside the tarfile.  As we're only
-          // parsing the file *name* as a GraphId here, we will just silently skip
-          // any file paths that can't be parsed by GraphId::GetTileId()
-          // If we end up with *no* recognizable tile files in the tarball at all,
-          // checks lower down will warn on that.
+      // map files to graph ids: use precomputed index.bin if available, else scan the tar
+      std::ifstream fin(*extract_fp, std::ios::binary);
+
+      tar::header_t first_header;
+      fin.read((char*)&first_header, sizeof(tar::header_t));
+      if (first_header.verify() &&
+          std::strstr(const_cast<char*>(first_header.name), "valhalla_tiles/index.bin")) {
+        // auto position = sizeof(tar::header_t);
+        fin.seekg(sizeof(tar::header_t));
+        auto position = fin.tellg();
+        while (fin.tellg() != first_header.get_file_size() + sizeof(tar::header_t)) {
+          uint32_t tile_id = 0;
+          uint64_t padded_tile_size = 0;
+          fin.read(reinterpret_cast<char*>(&tile_id), sizeof(tile_id));
+          fin.read(reinterpret_cast<char*>(&padded_tile_size), sizeof(padded_tile_size));
+
+          tiles[tile_id] = std::make_pair(reinterpret_cast<char*>(&position), padded_tile_size);
+
+          position += padded_tile_size;
         }
-      }
-      // couldn't load it
-      if (tiles.empty()) {
-        LOG_WARN("Tile extract contained no usuable tiles");
-      } // loaded ok but with possibly bad blocks
-      else {
-        LOG_INFO("Tile extract successfully loaded with tile count: " + std::to_string(tiles.size()));
-        if (archive->corrupt_blocks) {
-          LOG_WARN("Tile extract had " + std::to_string(archive->corrupt_blocks) + " corrupt blocks");
```

Comment:
hard to see, but really only shuffling around the conditions, nothing changed actually in this bit